examples: add TensorFlow to requirements.txt

JohannesGaessler opened the pull request:
I followed the instructions in `examples/mnist` and got the following error:

```
(ggml_env) 
/home/johannesg/Projects/ggml/examples/mnist [git::master *] [johannesg@johannes-pc] [20:16]
> py mnist-cnn.py                        
Traceback (most recent call last):
  File "/home/johannesg/Projects/ggml/examples/mnist/mnist-cnn.py", line 5, in <module>
    from tensorflow import keras
ModuleNotFoundError: No module named 'tensorflow'
```

I was able to fix it by explicitly adding TensorFlow to the requirements. The version that pip installs by default then also required bumping the Keras version. I added small filename-related changes to `examples/mnist/mnist-cnn.py` because otherwise there would be an error when using the exact instructions in the README (don't know whether that was already an issue with an older Keras version). Other than that MNIST and GPT-2 seem to work correctly. From my understanding the only other example that uses TensorFlow and Keras is Magika. However, for this example the model conversion fails:

```
2024-07-28 20:38:03.095525: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-07-28 20:38:03.105424: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-07-28 20:38:03.108430: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-07-28 20:38:03.116240: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-07-28 20:38:03.739007: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
Traceback (most recent call last):
  File "/home/johannesg/Projects/ggml/examples/magika/convert.py", line 32, in <module>
    convert(model_file)
  File "/home/johannesg/Projects/ggml/examples/magika/convert.py", line 6, in convert
    model = keras.models.load_model(model_name, compile=False)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/saving/saving_api.py", line 189, in load_model
    return legacy_h5_format.load_model_from_hdf5(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/legacy/saving/legacy_h5_format.py", line 133, in load_model_from_hdf5
    model = saving_utils.model_from_config(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/legacy/saving/saving_utils.py", line 85, in model_from_config
    return serialization.deserialize_keras_object(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/legacy/saving/serialization.py", line 495, in deserialize_keras_object
    deserialized_obj = cls.from_config(
                       ^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/models/model.py", line 521, in from_config
    return functional_from_config(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/models/functional.py", line 477, in functional_from_config
    process_layer(layer_data)
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/models/functional.py", line 457, in process_layer
    layer = saving_utils.model_from_config(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/legacy/saving/saving_utils.py", line 85, in model_from_config
    return serialization.deserialize_keras_object(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/legacy/saving/serialization.py", line 473, in deserialize_keras_object
    (cls, cls_config) = class_and_config_for_serialized_keras_object(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johannesg/Projects/ggml/ggml_env/lib/python3.12/site-packages/keras/src/legacy/saving/serialization.py", line 354, in class_and_config_for_serialized_keras_object
    raise ValueError(
ValueError: Unknown layer: 'TFOpLambda'. Please ensure you are using a `keras.utils.custom_object_scope` and that this object is included in the scope. See https://www.tensorflow.org/guide/keras/save_and_serialize#registering_the_custom_object for details.
```

This could be due to the change in Keras version since it already fails when just loading the model. Can someone for which the example is working please share their Python, Keras, and TensorFlow version?

slaren commented:
It works for me with these:
```
keras                     2.15.0             pyhd8ed1ab_0    conda-forge
python                    3.10.13         hd12c33a_1_cpython    conda-forge
tensorflow                2.15.0          cpu_py310h7825f03_2    conda-forge
```

JohannesGaessler commented:
Thank you, the problem was the Python version. I was using Python 3.12 but the venv needs to be >= 3.7 and <= 3.10. The build instructions were not present for all examples and they were inconsistent; I moved them to the main README and specified Python 3.10 there. Even with Python 3.10 TensorFlow was not getting installed automatically on my system so I added it to requirements.txt; with those versions the examples now run on my system. While I was going through the examples I also added bash syntax highlighting to relevant code blocks. The changes I made to `examples/mnist/mnist-cnn.py` seem to now no longer be necessary but I would suggest we keep them since that version of the code seems to work with both Keras 2 and Keras 3.
